lib/mongo/server/connection_pool/available_stack.rb

```diff
@@ -283,48 +263,92 @@ def max_idle_time
+          mutex.synchronize do
+            to_refresh = []
+            connections.reject! do |connection|
+              if last_checkin = connection.last_checkin
+                if (Time.now - last_checkin) > max_idle_time
+                  to_refresh << connection
+                end
               end
             end
-          end
-
-          mutex.synchronize do
-            num_checked_out = pool_size - queue_size
-            min_size_delta = [(min_size - num_checked_out), 0].max
 
             to_refresh.each do |connection|
-              if queue.include?(connection)
-                connection.disconnect!
-                if queue.index(connection) < min_size_delta
-                  begin; connection.connect!; rescue; end
-                end
-              end
+              close_connection(connection, Monitoring::Event::Cmap::ConnectionClosed::STALE)
             end
           end
         ensure
           check_count_invariants
         end
 
+        def connection_removed
+          @pool_size -= 1
+          if @pool_size < 0
+            # This should never happen
+            log_warn("ConnectionPool::AvailableStack: unexpected connection removal")
+            @pool_size = 0
+          end
+        end
+
         private
 
-        def dequeue_connection
-          deadline = Time.now + wait_timeout
+        def close_connection(connection, reason)
+          connection_removed
+          connection.disconnect!
+
+          publish_cmap_event(
+            Monitoring::Event::Cmap::ConnectionClosed.new(
+              reason,
+              @address,
+              connection.id,
+            )
+          )
+        end
+
+        def close_if_stale!(connection)
+          if connection.generation != @generation
+            close_connection(connection, Monitoring::Event::Cmap::ConnectionClosed::STALE)
+            true
+          end
+        end
+
+        def close_if_idle!(connection)
+          if connection && connection.last_checkin && max_idle_time
+            if Time.now - connection.last_checkin > max_idle_time
+              close_connection(connection, Monitoring::Event::Cmap::ConnectionClosed::IDLE)
+              true
+            end
+          end
+        end
+
+        def pop_connection(deadline)
+          mutex.synchronize do
+            get_connection(deadline)
+          end
+        end
+
+        def get_connection(deadline)
           loop do
-            return queue.shift unless queue.empty?
+            until connections.empty?
+              connection = connections.shift
+              return connection unless close_if_stale!(connection) || close_if_idle!(connection)
+            end
+
             connection = create_connection
             return connection if connection
-            wait_for_next!(deadline)
+
+            wait = deadline - Time.now
+            connection_available_condvar.wait(mutex, wait)
+
+            # TODO: Raise different error
```

Comment:
The consensus in the pull request for the spec change to add a separate error seemed to be against it. I think we can still fix this on our end though; if we make a subclass of `WaitQueueTimeout`, we can raise that here instead, which will let us put the info we want in the error message while still fulfilling the spec requirement that `WaitQueueTimeout` is raised, and users can still just catch `WaitQueueTimeout` if they don't care about which type is raised.

Reply:
The base exception class should be called something like `ConnectionCheckoutError`, with the `WaitQueueTimeoutError` being its subclass.

Reply:
And let's drop "Error" from `WaitQueueTimeoutError`, the only two classes with Error at the end is SocketError and SocketTimeoutError and the latter of them can also be spelled without the Error.